### PR TITLE
Add Float16 and BFloat16 support to C# API

### PR DESCRIPTION
**Pull request by yuslepukhin**
**Description**:
Address issue https://github.com/microsoft/onnxruntime/issues/948
Provide a way to feed/fetch data to evaluate ONNX model using onnxruntime.

**Motivation and Context**
People want to fine prune their models and save space.

### CODE REVIEW

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs [hunk 1]
```diff
@@ -1520,20 +1520,47 @@ private void TestModelInputUINT64()
             }
         }
 
-        [Fact(Skip = "FLOAT16 not available in C#")]
+        [Fact]
```
Comment (hariharans29) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs:1523]: I think there are full model tests that can be removed from exclusion as well

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs [hunk 2]
```diff
@@ -1520,20 +1520,47 @@ private void TestModelInputUINT64()
...
         {
             // model takes 1x5 input of fixed type, echoes back
-            string modelPath = Path.Combine(Directory.GetCurrentDirectory(), "test_types_FLOAT16.pb");
+            string modelPath = Path.Combine(Directory.GetCurrentDirectory(), "test_types_FLOAT16.onnx");
             using (var session = new InferenceSession(modelPath))
             {
                 var container = new List<NamedOnnxValue>();
-                var tensorIn = new DenseTensor<float>(new float[] { 1.0f, 2.0f, -3.0f, float.MinValue, float.MaxValue }, new int[] { 1, 5 });
+                var tensorIn = new DenseTensor<Float16>(
+                    new Float16[] { 15360, 16384, 16896, 17408, 17664 }, new int[] { 1, 5 });
```
Comment (hariharans29) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs:1532]: So the users still need to use third party libraries to obtain the corresponding ushort values from their half precision floats ? I thought that the added structs would have capability to arrive at the ushort values (half precision float) from a given single precision / double precision float

**Comment by hariharans29** [comment]
Can you reference the GH issue that this is solving so that they have a chance to review and ask questions ?

File: csharp/testdata/test_input_FLOAT16.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import onnx
```
Comment (pranavsharma) [inline csharp/testdata/test_input_FLOAT16.py:1]: License header is missing.

File: csharp/testdata/test_input_BFLOAT16.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import onnx
```
Comment (pranavsharma) [inline csharp/testdata/test_input_BFLOAT16.py:1]: License header.

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs (hunk 2 above)
Comment (yuslepukhin) [reply csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs:1532]: None of our APIs provide any conversions.

---
In reply to: [521762473](https://github.com/microsoft/onnxruntime/pull/5775#discussion_r521762473) [](ancestors = 521762473)

File: csharp/src/Microsoft.ML.OnnxRuntime/DisposableNamedOnnxValue.cs [hunk 5]
```diff
@@ -55,15 +54,28 @@ public void Dispose()
...
         private bool _disposed = false;
 
+        /// <summary>
+        /// Ctor
+        /// </summary>
+        /// <param name="name">Name of the output value</param>
+        /// <param name="value">Managed object created to represent output value, such as DenseTensor<T>
+        /// List or Dictionary
+        /// </param>
+        /// <param name="onnxValueType">Use this to decided what you want to call to fetch data, AsTensor(), AsDictionary()
```
Comment (pranavsharma) [inline csharp/src/Microsoft.ML.OnnxRuntime/DisposableNamedOnnxValue.cs:73]: nit: typo: decide*

**Review by pranavsharma — APPROVED** [review]

**Comment by tianleiwu** [comment]
Shall we support Half data type in .Net 5.0 (which is officially released)?
https://devblogs.microsoft.com/dotnet/introducing-the-half-type/
It can cast explicitly from/to float or double. That will be helpful.

**Comment by tianleiwu** [comment]
Do we have plan to support half data type in [.Net 5.0](https://dotnet.microsoft.com/download/dotnet/5.0)?
https://devblogs.microsoft.com/dotnet/introducing-the-half-type/
It can cast explicitly from/to float or double. That will be helpful.

**Review by RyanUnderhill — APPROVED** [review]

**Pull request merged**